### PR TITLE
fix(settlement): write reverse job mapping and harden startup recovery

### PR DESCRIPTION
**Pull request by hadjiszs**
Part of #1609. Settlement-job persistence and recovery hardening from the #1393 review ([reverse CF](https://github.com/agglayer/agglayer/pull/1393#discussion_r3519527758), [boot resilience](https://github.com/agglayer/agglayer/pull/1393#discussion_r3519527752)), plus the interrupted-completion TODO folded in from #1577.

**Reverse mapping.** `insert_certificate_settlement_job_id` now writes the certificate-to-job and job-to-certificate mappings in one atomic batch. The settlement storage only exists on test networks so far; jobs written before this change simply lack the reverse link, and nothing reads it yet, so no migration is added.

**Boot recovery.** One unreadable settlement job no longer prevents the node from booting. Startup recovery skips it, logs an error, and counts it in a new `agglayer_node_settlement_recovery_skipped_jobs` metric, so the job can be found and repaired. The count is recorded from `agglayer-node` to keep the service crate's dependency graph unchanged (the isolated feature checks are sensitive to it). A failure to scan the store as a whole still stops the boot. A skipped job that is queried later fails loudly to its certificate task.

**Interrupted completion writes.** A crash between the attempt-result writes and the final job-result write leaves the job looking pending. On restart the run loop re-derives every nonce's outcome from L1 rather than trusting stored state, so the interrupted completion is eventually finished through the loop's normal checks (inclusion, reorg detection, confirmation wait) regardless of the order nonces are scanned in. A still-pending nonce may bump or rebroadcast before the loop reaches the confirmed one, but no completion path bypasses those checks. Interrupted revert completions re-derive from L1 the same way.

Tests: reverse-link storage tests, skip-and-continue startup recovery, interrupted-completion recovery test, metric export test.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/agglayer-settlement-service/src/settlement_service.rs [hunk 1]
```diff
@@ -117,25 +128,33 @@ impl<
...
+                // An unloadable job must not prevent node boot: skip it and
+                // report loudly so it can be inspected and repaired.
+                Err(error) => {
+                    error!(
+                        ?error,
+                        %job_id,
+                        "Failed to load settlement job during startup recovery; skipping"
+                    );
+                    skipped_jobs += 1;
+                }
```
Comment (Freyskeyd) [inline crates/agglayer-settlement-service/src/settlement_service.rs]: In which case are we falling into this branch?

File: crates/agglayer-settlement-service/src/settlement_task.rs [hunk 2]
```diff
@@ -523,6 +523,14 @@ impl<
                 return run_result;
             }
 
+            // A recorded successful attempt result means a previous run
+            // already confirmed settlement per the configured policy but
+            // stopped before finishing the completion writes; finish them
+            // before any submission logic can run.
+            if let Some(job_result) = self.try_complete_interrupted_success().await {
+                return SettlementTaskRunResult::Completed(job_result);
```
[the remark this answers is not in the document]
Comment (hadjiszs) [reply crates/agglayer-settlement-service/src/settlement_task.rs]: Reworked in d04509154. The special resume path is gone. The run loop now simply processes the nonce that has a recorded success first, and completion always goes through the loop's normal checks: transaction on L1, reorg check, confirmation wait. A stale stored receipt can no longer complete a job, because no code completes a job from a stored receipt anymore.

File: crates/agglayer-settlement-service/src/settlement_service.rs (hunk 1 above)
Comment (hadjiszs) [reply crates/agglayer-settlement-service/src/settlement_service.rs]: This branch runs when a settlement job cannot be read back from the database at startup: the row is corrupt or cannot be decoded. This should never happen in normal operation; realistic causes are a damaged database or manual database edits.

Before this PR, one broken row prevented the whole node from booting. Now the node boots, skips that job, logs an error, and counts it in the `agglayer_node_settlement_recovery_skipped_jobs` metric so it can be found and repaired. This is also explained in a comment on the branch now.

File: crates/agglayer-settlement-service/src/settlement_task.rs [hunk 3]
```diff
@@ -1692,17 +1692,40 @@ impl<
...
+    /// successful attempt result come first.
+    ///
+    /// Such a result is only written after settlement-policy confirmation,
+    /// so at most the terminal job-result write is missing (a crash between
+    /// the completion writes). Handling that nonce first lets the loop finish
+    /// the interrupted completion through its normal L1 checks before any
+    /// other nonce can submit a new transaction; re-recorded completion
+    /// writes are idempotent no-ops. Interrupted revert completions need no
+    /// ordering: they re-derive from L1 wherever the loop starts.
+    fn nonces_in_processing_order(&self) -> Vec<(Address, Nonce)> {
```
Comment (Ekleog-Polygon) [inline crates/agglayer-settlement-service/src/settlement_task.rs]: **P1 simplification: drop the ordering — the loop already finishes interrupted completions unordered, in the same pass**

Nothing processed before the winning nonce can complete the job or block the loop, so the winner early-returns `Completed` later in the same pass even without this ordering. All the ordering buys is that other pending nonces may skip one extra gas bump each, and only when the process died exactly between the attempt-result write and the job-result write. That corner case doesn't justify a dedicated function plus the order-sensitive test machinery.

Concretely: iterate `all_used_nonces()` as before, delete this function and its unit test, and replace the removed TODO at `write_job_result_to_db` with a one-line note that the normal run-loop re-derivation finishes interrupted completions. Keep a slimmed `run_finishes_interrupted_completion_before_other_nonces` (mock both nonces, don't assert order) — that's the real regression guard for deleting the TODO.

File: crates/agglayer-storage/src/stores/interfaces/writer.rs [hunk 4]
```diff
@@ -70,7 +70,8 @@ pub trait StateWriter: Send + Sync {
 
     fn remove_settlement_tx_hash(&self, certificate_id: &CertificateId) -> Result<(), Error>;
 
-    /// Inserts the settlement job id associated with `certificate_id`.
+    /// Inserts the settlement job id associated with `certificate_id`,
+    /// writing the certificate→job and job→certificate mappings atomically.
```
Comment (Ekleog-Polygon) [inline crates/agglayer-storage/src/stores/interfaces/writer.rs:74]: **P1 followup: "Closes #1609" also closes item 3, which this PR doesn't implement**

Issue #1609 item 3 (recover dangling certificate→job links when the crash lands between this link write and the job save) is untouched here: `list_settlement_job_ids` reads the jobs-CF keys, so a dangling link never enters startup recovery, and the sentence below this hunk — "startup recovery can recreate a missing job from the certificate-side id" — describes behavior that doesn't exist yet. Please split item 3 into its own issue before merging (or change `Closes` to `Part of`), and point this doc sentence at it so it reads as intent, not as implemented behavior.

File: crates/agglayer-settlement-service/src/settlement_service.rs [hunk 5]
```diff
@@ -42,6 +42,7 @@ pub struct SettlementService<L1Provider, SettlementStore> {
     /// concurrent settlement tasks.
     /// XREF: https://github.com/agglayer/agglayer/issues/1597
     wallet_nonce_locks: Arc<WalletNonceLocks>,
+    recovery_skipped_jobs: u64,
```
Comment (Ekleog-Polygon) [inline crates/agglayer-settlement-service/src/settlement_service.rs]: **P3 nit: return the count from `start()` instead of storing it on the service**

This is one-shot boot state carried forever in a `Clone`-able long-lived struct, read exactly once by `agglayer-node` right after `start()`. Having `start()` return `(Self, u64)` keeps boot statistics out of steady-state service state and drops the `let mut this` reassignment, the getter, and the duplicated doc line on `resume_pending_settlement_jobs`.

File: crates/agglayer-telemetry/src/settlement/tests.rs [hunk 6]
```diff
@@ -140,3 +140,28 @@ fn settlement_metrics_export_the_issue_1676_series() {
         Some(1.0),
     );
 }
+
+#[test]
+fn recovery_skipped_jobs_counter_exports() {
+    let harness = MetricsHarness::install();
+
+    record_settlement_recovery_skipped_jobs(0);
```
Comment (Ekleog-Polygon) [inline crates/agglayer-telemetry/src/settlement/tests.rs]: **P3 test: the zero-export claim is never actually asserted**

The helper's doc says "a zero count still exports the series" — the property that lets alerting distinguish "scan ran, nothing skipped" from "no data". But this test only asserts the sum after `add(0)` + `add(2)`, which passes even if `add(0)` alone exported nothing. Assert the series exists with value `0.0` after the first call (a second `harness.gather()` before the `add(2)`), so the claim is pinned.

**Review by Ekleog-Polygon — COMMENTED** [review]
Reviewed at a9e8afab (CI green).

The reverse-mapping write and the boot skip-and-log both look right and are well-tested, and I like that interrupted completions now resume through the run loop's normal L1 checks instead of trusting stored state.

Four comments, none blocking:
- **P1 simplification**: the `nonces_in_processing_order` mechanism can go entirely — the loop already finishes interrupted completions unordered, in the same pass.
- **P1 followup**: `Closes #1609` would also close item 3 (dangling cert→job link recovery), which this PR doesn't implement — please split it out first.
- **P3 nit**: return the recovery-skip count from `start()` instead of storing it on the service.
- **P3 test**: the "zero count still exports" claim isn't actually asserted.

File: crates/agglayer-settlement-service/src/settlement_task.rs (hunk 3 above)
Comment (hadjiszs) [reply crates/agglayer-settlement-service/src/settlement_task.rs]: Agreed, removed it. The loop iterates `all_used_nonces()` again, and the function + its unit test are gone. The old TODO at `write_job_result_to_db` is now a one-line note that the run loop re-derives interrupted completions from L1. Kept a slimmed `run_finishes_interrupted_completion_before_other_nonces`: the competing nonce is scanned first, doesn't block, then the winner completes — order no longer asserted.

File: crates/agglayer-storage/src/stores/interfaces/writer.rs (hunk 4 above)
Comment (hadjiszs) [reply crates/agglayer-storage/src/stores/interfaces/writer.rs:74]: You're right, item 3 (dangling cert→job link recovery) isn't implemented here. Reworded the doc so it reads as future work. Rather than split it out, I'll keep #1609 open and change this PR to `Part of #1609` — item 3 lands as a follow-up PR that closes it.

File: crates/agglayer-settlement-service/src/settlement_service.rs (hunk 5 above)
Comment (hadjiszs) [reply crates/agglayer-settlement-service/src/settlement_service.rs]: Done. `start()` now returns `(Self, u64)`; dropped the field, the getter, the `let mut this`, and the duplicated doc line. `agglayer-node` takes the count from the return value.

File: crates/agglayer-telemetry/src/settlement/tests.rs (hunk 6 above)
Comment (hadjiszs) [reply crates/agglayer-telemetry/src/settlement/tests.rs]: Good catch. The test now gathers right after `add(0)` and asserts the series is `0.0`, before `add(2)` — so the zero-export claim is actually pinned.

File: crates/agglayer-storage/src/stores/interfaces/writer.rs [hunk 7]
```diff
@@ -70,12 +70,14 @@ pub trait StateWriter: Send + Sync {
...
     ///
     /// This is an insert-only operation and must fail if `certificate_id`
     /// already has a stored settlement job id. The settlement job may be
-    /// created after this link, so startup recovery can recreate a missing job
-    /// from the certificate-side id.
+    /// created after this link; recovering a dangling link whose job save was
+    /// interrupted is intended future work, not implemented here (see #1609
```
Comment (Ekleog-Polygon) [inline crates/agglayer-storage/src/stores/interfaces/writer.rs:79]: P1 issue: Keep #1609 open for dangling-link recovery

Thanks, the documentation now correctly identifies item 3 as future work. The current PR description still starts with `Closes #1609`, though, so merging would close the issue while dangling certificate→job recovery remains unimplemented. Please change it to `Part of #1609`, as proposed in your reply.

File: crates/agglayer-settlement-service/src/settlement_task/tests.rs [hunk 8]
```diff
@@ -904,6 +904,188 @@ async fn write_nonce_revert_replaces_previous_client_error_for_finalized_attempt
...
+    }
+}
+
+#[tokio::test]
+async fn run_finishes_interrupted_completion_before_other_nonces() {
+    // Regression guard for dropping the resumption TODO: a completion
+    // interrupted between the attempt-result write and the job-result write
+    // (the winning nonce carries a recorded success, but the terminal job
+    // result never landed) is finished by the run loop re-deriving it from L1.
+    // A competing pending nonce is scanned first, yet nothing before the
```
Comment (Ekleog-Polygon) [inline crates/agglayer-settlement-service/src/settlement_task/tests.rs:999]: P2 doc: Update the interrupted-completion claim after removing ordering

The code now intentionally scans nonces in their normal order, and this test makes the earlier nonce’s wallet unknown so it cannot submit. With a known signer and an expired retry deadline, that nonce may still bump or rebroadcast before the recorded success is reached—the tradeoff accepted in the prior thread. The PR description still says successful nonces run first and no other nonce can submit anything; please update that paragraph to promise eventual completion through the normal L1 checks instead.

**Review by Ekleog-Polygon — DISMISSED** [review]
Reviewed at `19b87d0c` (CI green).

The previous code concerns are addressed: the dedicated recovery ordering is gone, the one-shot recovery count is returned instead of retained in service state, and zero-valued metric export is now tested. The full diff looks sound: reverse mappings are written atomically, unreadable per-job state no longer blocks node boot, and interrupted completions still pass through the normal live L1 checks.

Two PR-description corrections remain: keep #1609 open for dangling-link recovery, and remove the outdated claim that recorded successes are processed before any other nonce can submit.

No code blockers from me.

**Review by Ekleog-Polygon — APPROVED** [review]
